### PR TITLE
improve support for traditional batch jobs

### LINKED ISSUE #2959: traditional batch job submission interface
I thought there was an open issue on this already, but I could not find it. Please close this as a duplicate if someone finds a previous open issue.

A traditional RM "batch" job in Flux can be thought of as a `flux-start` of a new instance with the "batch script" as the initial program of the new instance. E.g., currently a more traditional looking batch job on 4 nodes would be submitted via

```
$ flux mini submit -N4 flux start /path/to/batch/script
```

We'll want an interface that at least makes the `flux start` implicit, allowing a user to just specify total resource requirement + their batch script.

Some other items of note (off the top of my head)

 * We may need support for "per resource" task slot assignment for this to work. By default, we'll want to allow the user to request any resources, but then run `flux-start` only once per node or execution target. Perhaps there is some other hack we could use here (like a "batch script" attribute on the jobspec) to make progress without a revision in jobspec.

 * There is not a very clear distinction between batch vs parallel job in Flux, as opposed to more traditional RMs like Slurm. I had thought Slurm could be configured such that partitions can be configured to only allow batch job submission (i.e. sbatch not srun), though I can't find evidence of that now. If there is some similar requirement we will have to be able to differentiate batch jobs in job-ingest.

 * Slurm creates a copy of batch scripts as part of the job record when using sbatch (at least as far as I remember). This means you can submit a script, edit it and submit again and have two different batch scripts submitted. I'm not sure if this is a requirement, but we might think about how a  Flux batch submission tool might support this. (i.e. include a batch script in jobspec itself)

### PR DESCRIPTION
**Pull request by grondo**
Throwing this PR up to start some discussion.

This PR adds support for traditional "batch script" submission to Flux via a new `flux mini batch` command. As discussed in #2959, this is meant to be a stopgap to cover a current gap in our tools, and may not be the final nor suggested interface for batch job submission in the long run.

Here's a summary of the changes so far:

 * Addition of a job-shell option `per-resource`, which if set must be a JSON object of the form `{type:s count?i}`, where type can be "node" or "core", and `count` is the number of tasks to launch on each resource `type`. This overrides the task slots from the jobpec.
 * Addition of a `batch` plugin to the job shell which looks for `attributes.system.batch.script`. If set, the rank 0 shell copies its contents to a temporary location and sets the executable bit, then rewrites task argv to be
 `flux broker [broker-opts] /path/to/script [args]`.  All other ranks rewrite task argv to
 `flux broker [broker-opts]`
 * If set `attributes.system.batch.broker-opts` is set, then this is used as `broker-opts` above.
 * Finally, a new `flux-mini batch` command is added, which checks to ensure the first non-option arg is a script and copies it into `attributes.system.batch.script`. It then sets the shell option `per-resource` to `{ "type": "node" }` so that only one task per node will be invoked by the shell.
 * To simulate traditional batch submission, `flux-mini batch` sets the default ouptut file to `flux-{{id}}.out`. This can be overridden with `--output` as expected, including `--output=none` to send output back to the kvs.

Some open questions:

 - The batch script and broker options just felt like they should be in `system.attributes.batch`, however, this could just as easily been kept as  a shell option instead.
 - Because of the way `flux-mini.py` is structured, as well as the `JobspecV1` class, the `flux mini  batch` command still takes basic options `-n, -N, -c, -g`. However, I didn't think that `--ntasks` or `--cores-per-task` made any sense, so they are renamed `--nslots`, `--cores-per-slot`, `--gpus-per-slot`. It occurs to me that perhaps `run` and `submit` could *also* take these alternate names if that is more user-friendly.
- The default output file of `flux-{{id}}.out` could get us into trouble since `id` is not guaranteed unique between many subinstances running at once. Maybe not a problem for now, but something to think about.

I'm pushing this up for comments before starting on any tests, in case there are objections to any interfaces presented here, but everything is basically working:

```
ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux mini batch -N2 -n4 test.sh
1601368489984
ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git$ cat flux-1601368489984.out
     STATE NNODES   NCORES    NGPUS
      free      2        4        0
 allocated      0        0        0
      down      0        0        0
9948889088
13002342400
             JOBID USER     NAME       STATUS NTASKS NNODES  RUNTIME RANKS
       13002342400 grondo   sleep           R      1      1   0.158s 0
        9948889088 grondo   hostname       CD      1      1   0.064s 0
2020-05-31T20:46:11.617796Z broker.err[0]: rc3: flux-job: Canceled 1 jobs (0 errors)

$ flux mini batch -N2 -n4 --broker-opts=-v test.sh
3183879389184
ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git$ cat flux-3183879389184.out 
flux-broker: boot: rank=1 size=2
flux-broker: parent: tcp://127.0.0.1:49152
flux-broker: boot: rank=0 size=2
flux-broker: parent: none
flux-broker: child: tcp://127.0.0.1:49152
flux-broker: initializing modules
flux-broker: installing session heartbeat: T=2.0s
flux-broker: loading connector-local
flux-broker: child: none
flux-broker: initializing overlay connect
flux-broker: initializing modules
flux-broker: loading connector-local
flux-broker: entering event loop
flux-broker: entering event loop
     STATE NNODES   NCORES    NGPUS
      free      2        4        0
 allocated      0        0        0
      down      0        0        0
9630121984
12918456320
             JOBID USER     NAME       STATUS NTASKS NNODES  RUNTIME RANKS
       12918456320 grondo   sleep           R      1      1   0.163s 0
        9630121984 grondo   hostname       CD      1      1   0.098s 0
2020-05-31T20:47:45.931657Z broker.err[0]: rc3: flux-job: Canceled 1 jobs (0 errors)
flux-broker: exited event loop
flux-broker: cleaning up
flux-broker: exited event loop
flux-broker: cleaning up

ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux mini batch -o verbose -N2 -n4 --output=none test.sh
4521275162624
ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux job attach 4521275162624 2>&1 | grep batch
0.078s: flux-shell[1]: DEBUG: batch: re-writing command to flux broker
0.068s: flux-shell[0]: DEBUG: output: batch timeout = 0.500s
0.076s: flux-shell[0]: DEBUG: batch: Copying batch script size=96 for job to /tmp/script-4521275162624-U3sVch
0.079s: flux-shell[0]: DEBUG: batch: re-writing command to flux broker /tmp/script-4521275162624-U3sVch

```

Fixes #2959

### CODE REVIEW

**Comment by garlick** [comment]
Wow neat.  So the batch script contents become part of the signed jobspec?  Nice!

**Comment by dongahn** [comment]
> The default output file of flux-{{id}}.out could get us into trouble since id is not guaranteed unique between many subinstances running at once. Maybe not a problem for now, but something to think about.

Just one (obvious) thought: If a nested instance has a way to join its local jobid with jobids of its ancestors, unique hierarchical id space can be used? Maybe a parent can push that ancestors jobid "prefix" when a nested instance is created...

**Comment by dongahn** [comment]
> Because of the way flux-mini.py is structured, as well as the JobspecV1 class, the flux mini batch command still takes basic options -n, -N, -c, -g. However, I didn't think that --ntasks or --cores-per-task made any sense, so they are renamed --nslots, --cores-per-slot, --gpus-per-slot. It occurs to me that perhaps run and submit could also take these alternate names if that is more user-friendly.

How would a user specify the shape of each slot with `flux-mini` interface? Is it determined implicitly?

**Comment by dongahn** [comment]
@grondo: I only have two comments for now. Very nice enhancements important for our testing on fluke. Thank you!

**Comment by grondo** [comment]
> How would a user specify the shape of each slot with flux-mini interface? Is it determined implicitly?

The slot shape is specified in the same way as other `flux-mini` subcommands, with `-g` and `-c`, only the long options in the case of `flux-mini batch` are (currently) `--gpus-per-slot=N` and `--cores-per-slot`.

> Just one (obvious) thought: If a nested instance has a way to join its local jobid with jobids of its ancestors, unique hierarchical id space can be used? 

Hm, good thought! We can have some nice long output filenames! For now, I wonder if probability of a FLUID collision is low enough for us to punt on a solution for now...

**Comment by grondo** [comment]
>  So the batch script contents become part of the signed jobspec?

Yeah, this worked out nicely, and made me think that in the future perhaps the configuration for child instances could be placed into jobspec as well.

Currently the script is simply written out to a unique name in `/tmp`. I wasn't sure if we had some *temporary*  per-broker directory into which this could go instead. There is `rundir`, but that may be persistent if specified by the user and the directory already exists, so I don't think this is what we want. For now, `TMPDIR` should work fine, but we may want a separate per-broker directory that *always* gets cleaned up after instance shutdown.

**Comment by grondo** [comment]
Pushed a fix for #2963.

**Comment by dongahn** [comment]
> Hm, good thought! We can have some nice long output filenames! For now, I wonder if probability of a FLUID collision is low enough for us to punt on a solution for now...

As a short term solution what you have should be fine. 

I just suggested this or similar as a long term solution. I think the hierarchy aware output file name has an advantage not only for collision avoidance but also debugging. FWIW, this trick worked pretty nicely for `flux-tree`.

**Comment by dongahn** [comment]
> The slot shape is specified in the same way as other flux-mini subcommands, with -g and -c, only the long options in the case of flux-mini batch are (currently) --gpus-per-slot=N and --cores-per-slot=N.

Ah. This makes sense and I agree `slot` would be better than `task`. How would you suggest we describe the concept of slot for users who are familiar with tasks. Probably by augmenting the help message for `flux-mini`?

**Comment by grondo** [comment]
> Probably by augmenting the help message for flux-mini?

That is a good thought. I will try to add something to `flux mini batch --help` that describes the concept of a slot. Hopefully it will be straightforward since a user will specify similar options as they'd use for `submit` or `run`.

**Comment by grondo** [comment]
If the proposed interface is largely acceptable, then I can work on some quick tests.

BTW, if this works out I wonder if we'll want a similar command, but for interactive use,
e.g. maybe a `flux mini alloc` that does the equivalent of `flux mini run -o pty [-n|-N|-c|-g] flux broker $SHELL`?

**Comment by dongahn** [comment]
> BTW, if this works out I wonder if we'll want a similar command, but for interactive use,
e.g. maybe a flux mini alloc that does the equivalent of flux mini run -o pty [-n|-N|-c|-g] flux broker $SHELL?

I like it!

**Comment by jameshcorbett** [comment]
I think this interface looks great.

**Comment by grondo** [comment]
ok, some basic functionality tests have been added. Further refactoring of `flux-mini.py` will be needed for a `flux mini alloc`.

I updated the description of `flux mini batch --nslots` in the `--help` output:
```
  -n, --nslots=N          Number of total resource slots requested. The size
                          of a resource slot may be specified via the -c,
                          --cores-per-slot and -g, --gpus-per-slot options.
                          The default slot size is 1 core.
```

Which should help explain the concept of slot and slot-size in the context of `mini batch`.

**Comment by grondo** [comment]
Ok, a little more incremental work here. Some very small amount of refactoring was done in `flux-mini.py` to allow for a `flux-mini alloc` subcommand. The interface is identical to `flux-mini batch` except the I/O options are suppressed, and the command takes optional `COMMAND` and args, not a script:

```
usage: flux mini alloc [COMMAND] [ARGS...]

Allocate resources and start a new instance. Once the instance is running
attach to it interactively.

positional arguments:
  COMMAND                 Set the initial COMMAND of new instance.(default is
                          the current shell)

optional arguments:
  -h, --help              show this help message and exit
  -t, --time-limit=FSD    Time limit in Flux standard duration, e.g. 2d, 1.5h
      --priority=N        Set job priority (0-31, default=16)
      --job-name=NAME     Set an optional name for job to NAME
  -o, --setopt=OPT        Set shell option OPT. An optional value is supported
                          with OPT=VAL (default VAL=1) (multiple use OK)
      --setattr=ATTR=VAL  Set job attribute ATTR to VAL (multiple use OK)
      --flags=FLAGS       Set comma separated list of job submission flags.
                          Possible flags: debug, waitable
      --dry-run           Don't actually submit job, just emit jobspec
  -N, --nodes=N           Distribute allocated resources across N individual
                          nodes
  -n, --nslots=N          Number of total resource slots requested. The size
                          of a resource slot may be specified via the -c,
                          --cores-per-slot and -g, --gpus-per-slot options.
                          The default slot size is 1 core.
  -c, --cores-per-slot=N  Number of cores to allocate per slot
  -g, --gpus-per-slot=N   Number of GPUs to allocate per slot
  -B, --broker-opts=OPT   Pass options to flux brokers

```

By default, `flux-mini alloc` just invokes `flux broker [broker-opts]` once per node on allocated resources. The `pty` shell option is used so that an interactive shell is spawned by default:

```
ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux mini alloc -n1
ƒ(s=1,d=1,builddir) grondo@asp:~/git/flux-core.git$ 
```
As with `flux start` or `flux broker` the default "initial command" can be overridden on the command line

```
ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux mini alloc -n5 flux resource list -v
     STATE NNODES   NCORES    NGPUS LIST
      free      2        5        0 rank0/core[0-3] rank1/core0
 allocated      0        0        0 
      down      0        0        0 
[detached: session exiting]
```


One other change in this latest round. For `flux mini batch` and `flux mini alloc` there is no default for `--nslots`, so this option at least must be specified. This is to prevent users being surprised when typing just `flux mini alloc` -- you get a warning message instead of an interactive shell to a Flux instance of just one core (since the shell is interactive, users might not even notice)

**Comment by jameshcorbett** [comment]
I'm not entirely sure what a "slot" is, @grondo, even with that help message. I'm not sure what the relationship is between `-N` and `-n` is, either. I saw in your example `flux mini batch -N2 -n4` gave you 2 nodes and 4 cores. I'm used to `-N` giving you access to whole nodes (except on the weird Slurm machines like Borax and rzansel)... does it not behave that way? 

Looking now at the help message for `-N` you just posted for `flux mini alloc`, it looks like `-N` behaves just like `-N` for `flux mini run` (i.e. distributing resources across nodes, not allocating entire nodes). If that's the case, I would probably suggest that there should be some option to express "give me all the resources on the nodes I requested with `-N`."

**Comment by grondo** [comment]
> "give me all the resources on the nodes I requested with -N."

Good feedback @jameshcorbett. Unfortunately, I don't think Flux has a way to do that right now. The `mini alloc` and `mini batch` subcommands currently behave exactly the same as `mini submit` and `mini run`. Since a "number of tasks" doesn't make sense in this context though, I chose to rename the options, but otherwise they behave exactly the same as the other "mini" commands.

I guess I should say technically jobspec does support requesting exclusive access to resources, however I'm not sure this is the case for V1 (and sched-simple certainly doesn't support it)

**Comment by grondo** [comment]
> I'm not entirely sure what a "slot" is, @grondo, even with that help message.

A slot is just a unit of resource, currently a count of cores and/or gpus, meant to be replicated multiple times with `-n, --nslots`. The name is arbitrary here, but is meant to reflective of the underlying "task slot" concept in jobspec. We can choose another term here that makes more sense, or just stick with `--ntasks` if that ends up being less confusing.

**Comment by jameshcorbett** [comment]
It's starting to make more sense, thanks. I see some use for fooling around with the slot options, because they let you request just a fraction of a node.

I'm not really sold on the use of `-n` though. When would you ever want it to be different from `-N`? E.g. if you have two slots per node, why not just double your slot size? (And I can't imagine a use case for, say, 5 slots on 4 nodes.)

I would propose, then, suppressing `-n` and changing the description to `-c` and `-g` to `-per-node`, not `-per-slot`. I think that would be clearer for users like me not familiar with slots. I'm assuming, though, that `-n` defaults to the value of `-N` if `-n` isn't set.

To address your concern about `flux mini alloc` giving you a interactive instance with just one core, you could make `-c` mandatory instead of `-n`.

**Comment by grondo** [comment]
> I would propose, then, suppressing -n and changing the description to -c and -g to -per-node, not -per-slot. I think that would be clearer for users like me not familiar with slots.

Ah, this is good feedback. I have to think if our current jobspec supports it, though.

>  I'm assuming, though, that -n defaults to the value of -N if -n isn't set.

No, not currently, though an error is produced if `-n` is not greater than or equal to `-N`.

Flux is not a node-centric resource manager like Slurm, so it isn't quite as straightforward to just ask for a certain number of nodes. In the future, I think we'll have much better support for this in the "real" front-end tools like `flux-run`.

The idea of a "task slot" currently is to allocate enough resources for a certain number of tasks which each require a certain number of cores and GPUs, without worrying about how many nodes those resources need to be spread across. E.g. this supports usage on clusters with heterogeneous nodes.

In the case of the `flux-mini` commands, the "slots" based interface is also motivated upon the underlying JobspecV1 class, as well as keeping the amount of code differences between different `flux mini` subcommands to a mininum.

All that being said, I agree that the concept of resource slots is confusing to users, so I'm willing to iterate on the interface here. I'm not sure if we can support your suggestion of dropping `-n` though. I think I might need @SteVwonder's help here.

**Comment by SteVwonder** [comment]
Thanks @jameshcorbett.  Excellent feedback!

tl;dr: maybe we keep `-n` for compatibility with `mini run`/`mini submit` but we support users eliding `-n` to make the common HPC use-case easier.

> All that being said, I agree that the concept of resource slots is confusing to users, so I'm willing to iterate on the interface here. I'm not sure if we can support your suggestion of dropping -n though.

I agree that the slot is confusing in the context of `alloc`/`batch` where typical HPC users aren't asking for lots of little slots to run a certain number of tasks within.  They are usually asking for a big bag of resources, which they will subdivide later and run tasks within using `mini run`/`mini submit`.  I normally just think of "I want X nodes" or "I want Y cores", but as @grondo pointed out on our coffee call, the nodes might not be homogenous.  If you are running in the cloud, each VM might have a different number of cores, and in that environment (where the spell of node-centricity is broken), the slot starts to make sense again.  For example, maybe the smallest resource unit (slot) that I can work with is 1 cores and 1 GPU (because I'm spinning up python processes which will use each use some ML lib to run on the GPU), and I want 36 of those slots (b/c I need a certain amount of GPUs to get the ML training done in time) regardless of the number of nodes/VMs that requires.

In terms of dropping `-n` from `flux mini alloc/batch`, I think the main downside of that would be that you can no longer take an arbitrary `flux mini run` command and substitute `alloc` for `run` and expect it to work.  You will have to reformulate any commands where `-N` != `-n` to get the right semantics.

> > I'm assuming, though, that -n defaults to the value of -N if -n isn't set.

> No, not currently, though an error is produced if -n is not greater than or equal to -N.

Maybe that is something we allow in `flux mini batch`/`alloc` to make it easier for users?  For example, `flux mini alloc -N 6 -c 18` would not error out and would get you 6 nodes each with 18 cores.  This seems kinda reasonable for these two sub-sub-commands since they result in a single task per node (by default) anyways.

**Comment by grondo** [comment]
Thanks, @SteVwonder, great suggestion.

I tend to think of resource slots a lot like the libc `calloc(3)` call. You want to allocate space for N things of size M, where M=1 core by default. In the case of `flux mini run/submit`, you also happen to be assigning 1 tasks to each of these N things, but in the case of `batch` or `alloc` you are allocating these things to be filled in later by a new Flux instance. :wink:

**Comment by jameshcorbett** [comment]
Ah, right, I forgot that you could use `-n`, `-c`, `-g` without `-N`, because it's a variation of the `mini` utilities. I was assuming that you always need to set `-N` or (if you don't) it defaults to 1, since that's how `sbatch` and `bsub` work.

Given what you have both said, I think the interface is pretty good as it is. I made issue #2969 to request that the `-N`/`-n` behavior becomes standard across the `flux mini` utility. (Why make it unique to `mini batch` and `mini alloc`?)

To make things easy for HPC users in the future, I think it might be nice to make a different utility (maybe something like `flux node-alloc` and `flux node-batch`) that's more node-centric and doesn't accept `-n`, `-c`, or `-g` options. But that can be a long way off.

> In terms of dropping -n from flux mini alloc/batch, I think the main downside of that would be that you can no longer take an arbitrary flux mini run command and substitute alloc for run and expect it to work. You will have to reformulate any commands where -N != -n to get the right semantics.

I thought that already wasn't possible? E.g. `flux mini alloc -N1 -n2 script.sh` behaves differently from `flux min run -N1 -n2 flux start script.sh` because the second will create two brokers on the node (possibly causing a shared-memory error) but the first won't.

**Comment by grondo** [comment]
> I thought that already wasn't possible? E.g. flux mini alloc -N1 -n2 script.sh behaves differently from flux min run -N1 -n2 flux start script.sh because the second will create two brokers on the node (possibly causing a shared-memory error) but the first won't.

That's true. However, in terms of resources allocated they behave the same. In the case of `flux mini alloc` and `batch` the idea is to get access to a set of resources with a new instance of Flux. As with the system instance, the end user should be oblivious of where brokers are actually running. So alloc and batch are interfaces for allocating resources (and subsequently running on them via a script or interactive session), while run and submit are interfaces for running parallel jobs on these resources.

I think you are completely correct that in the future we'll need better tools!

**Comment by dongahn** [comment]
JWIW, slot is a concept being used widely in the scheduling domain so as far as we explain this thorough to our users, I think we should be okay...

> In terms of dropping -n from flux mini alloc/batch, I think the main downside of that would be that you can no longer take an arbitrary flux mini run command and substitute alloc for run and expect it to work. You will have to reformulate any commands where -N != -n to get the right semantics.

I think this is kind of important. But to make sure, I want to understand the current semantics though. If you had

`flux mini run -N 2 -n 4 foo` and relace it with

`flux mini batch -N 2 -n 4 --wrap flux mini run foo`

Will `foo` be run across 2 nodes each with 4 tasks? I think this one level inheritance makes sense.

Now, if you instead have

`flux mini batch -N 2 -n 4 --wrap flux mini batch ./myscript.sh`

`-n 4` should affect neither the nested flux mini batch command nor my script correct? Since multiple inheritance will make things too confusing?

**Comment by dongahn** [comment]
This is not a part of this PR. But now that I think about this, currently, the biggest problem with Flux's hierarchical composability arises when a WF tool wants to treat MPI programs and flux-start pretty much equally. Consider, the tool gets a task list which can consist of both mph programs and nested flux.  

`flux submit -N 2 -n 4 mpi_prog`

vs.

`flux submit -N 2 -n 4 flux start ...`


The latter *almost* worked, but when we run more than one broker runs on each node, we found this creates shared memory problems for nested MPI programs.

`flux-tree` solves this its own way, but I wonder if this nesting composability problem can be solved more genetically at the flux-score level.

**Comment by grondo** [comment]
> `flux mini batch -N 2 -n 4 --wrap flux mini run foo`
> Will foo be run across 2 nodes each with 4 tasks? I think this one level inheritance makes sense.

No, there is no inheritance, and I don't think there is a clean way to do it either. I think the above `flux mini batch` command should just be:

```
flux mini submit -N2 -n4 foo
```

To avoid the useless launch of a new instance.


> `flux mini batch -N 2 -n 4 --wrap flux mini batch ./myscript.sh`
> -n 4 should affect neither the nested flux mini batch command nor my script correct? Since multiple inheritance will make things too confusing?

I honestly think trying to implement this inheritance will be fraught with peril.

I think instead of doing this automatically, it should be opt-in only, by making the necessary values discoverable via scriptable commands.

**Comment by grondo** [comment]
>Flux's hierarchical composability arises when a WF tool wants to treat MPI programs and flux-start pretty much equally. 

We perhaps should discuss this in a new issue. I don't follow why a workflow tool would want to treat MPI programs the same as a new Flux instance. One can run more programs and the other cannot, so it seems they would have to be treated differently somehow.

If, however, this is really a requirement for some use case, then I would think a higher level utility would be warranted, which offers the interface required. E.g. you could write a script that passes through args to `flux mini submit` unless the string "flux start" appears in the commandline, in which case it passes through to `flux mini batch`.

**Comment by dongahn** [comment]
> No, there is no inheritance, and I don't think there is a clean way to do it either. I think the above flux mini batch command should just be:

Ah... So `-n` is just needed as a convenience for users to specify their resource requirements and shape then. Now I understand why you change the long option name from `-ntasks` to `-nslots`. IBM RM introduce the concept of `resource set` but that confuses our users whole lot.

I'm okay with `-nslots` for `flux mini batch`.

**Comment by grondo** [comment]
> Ah... So -n is just needed as a convenience for users to specify their resource requirements and shape then. Now I understand why you change the long option name from -ntasks to -nslots.

Yes! It allows you to specify resource requirements "as if" you were running tasks. I wonder if there is a better way to document it though? I think of it like "Create me a Flux instance with enough resources to run 64 tasks each requiring 4 cores and GPU": `flux mini batch -n64 -c 4 -g 1 myscript.sh`

**Comment by dongahn** [comment]
> We perhaps should discuss this in a new issue. I don't follow why a workflow tool would want to treat MPI programs the same as a new Flux instance. One can run more programs and the other cannot, so it seems they would have to be treated differently somehow.

When you have a chance (not a high priority though), could you review Brett Sibert's (from Naval Nuclear Lab) case? I thought he expressed one of the common desires that workflow users have. The initial approach was same for @jameshcorbett and we addressed his needs with `flux-tree`. But there would be users `flux-tree` isn't a good fit. 

BTW, I'm not saying this is a high priority.

**Comment by dongahn** [comment]
> Yes! It allows you to specify resource requirements "as if" you were running tasks. I wonder if there is a better way to document it though? I think of it like "Create me a Flux instance with enough resources to run 64 tasks each requiring 4 cores and GPU": flux mini batch -n64 -c 4 -g 1 myscript.sh

IBM folks introduced the concept of resource set to express a similar thing. But honestly I like slots better than resource set...

> I wonder if there is a better way to document it though?

Our new flux-docs and bubble this up to a high level enough page?

**Comment by garlick** [comment]
One thought about usability: if alloc/batch define the allocation in terms of slots, is there any way to request one of _those_ slots from within the allocation?  E.g. alloc/batch N slots that each contain a GPU and four cores and then `flux mini run -n4 cmd` and expect to get 4 tasks on 16 cores and 4 GPUs, instead of 4 tasks on 4 cores?

it's maybe a variant on the inheritance question so sorry if I failed to detect if the question has  already been asked and answered.

**Comment by grondo** [comment]
> Maybe that is something we allow in flux mini batch/alloc to make it easier for users? For example, flux mini alloc -N 6 -c 18 would not error out and would get you 6 nodes each with 18 cores. This seems kinda reasonable for these two sub-sub-commands since they result in a single task per node (by default) anyways.

I was going to implement this suggestion but when thinking about testing it, I came to the conclusion that it may not currently be a good idea, at least until we have an implementation that can ask for whole nodes. E.g. I don't think this should work by default
```
flux mini batch -N2 script.sh
```

Because currently that gets you 1 core on each of two nodes, which I think would be surprising. Especially, as @jameshcorbett pointed out, this is not how any current tools work.

Once we have a way to allocate whole nodes exclusively, we could revisit this. (and maybe we already do and I just don't know it.) Even then

For now I'll leave the current behavior that `-n, --nslots` is a required argument. This hints to the user that they are forming their resource request in terms of slots, not in terms of whole nodes, which I fear the behavior of defaulting `-n` to `-N` would encourage.

**Comment by grondo** [comment]
> One thought about usability: if alloc/batch define the allocation in terms of slots, is there any way to request one of those slots from within the allocation? E.g. alloc/batch N slots that each contain a GPU and four cores and then flux mini run -n4 cmd and expect to get 4 tasks on 16 cores and 4 GPUs, instead of 4 tasks on 4 cores?

This is a good suggestion, though I think users would ask why not *also* default the number of slots such that `flux mini run cmd` "works", and we're back to magic environment variables.

Maybe I'm just being overly cautious by shying away from the use of environment variables to set defaults for all `flux-mini` options?  I would feel better if it was opt-in (sorry I sound like a broken record), e.g. either via a new flux-mini option `--from-env`, or an environment setup option `flux mini setenv` (sorry can't think of better names ATM.)

**Comment by garlick** [comment]
I was thinking more along the lines of an instance having a default slot shape so that jobspec would not fully define the slots?  As I type that now I'm realizing that's probably not allowed in jobspec.

**Comment by grondo** [comment]
> I was thinking more along the lines of an instance having a default slot shape so that jobspec would not fully define the slots? As I type that now I'm realizing that's probably not allowed in jobspec.

I'm not sure, currently, how the default slot shape would be stored and communicated to various entities in the instance.

However, I think this idea could be a really powerful concept, and could maybe even be a huge boon to scheduler scalability! If a user could create an instance and tell it the schedulable unit, then the child scheduler could just schedule at this granularity instead of cores, and why not call this a "slot" resource? (well maybe overloading that term is a bad idea, but maybe a "vslot" or something)

**Comment by grondo** [comment]
> I'm not sure, currently, how the default slot shape would be stored and communicated to various entities in the instance.

Well, I can think of two ways, neither of which I like

 * Export environment variable to the job `FLUX_SLOT_SHAPE=core:2,gpu:1` or similar. I don't like this because then it is an environment variable you have to "clean" whenever that default needs to be reset, and inevitably this gets missed somewhere
 * Store defaults in the kvs somewhere, and fetch them from commands. This requires an extra kvs access for every command that needs to read the defaults.

**Comment by garlick** [comment]
I hate to get this PR off track - is this idea worthy of a separate issue?

**Comment by grondo** [comment]
> I hate to get this PR off track - is this idea worthy of a separate issue?

I can't tell if so-called "inheritance" is a requirement here or not. I don't see it as such, but not sure how others feel (based on comments it seems to be important)

**Comment by garlick** [comment]
I'm fine with assuming it's not a requirement for now and proceeding with `mini batch` as designed.

**Comment by grondo** [comment]
Another idea, since our front-end tools are python, we could trivially support template substitution in job scripts.

**Comment by grondo** [comment]
BTW, what I'd personally like to avoid is this:
```
$ sbatch -N1 -p pdebug
salloc: Granted job allocation 5352443
$ env | grep -c SLURM
$ env | grep -c ^SLURM
47
```
:cry:
```

**Comment by grondo** [comment]
Just forced a push with some refactoring and squashing. There are some basic functionality tests for `flux mini alloc` now, and a `--wrap` option has been added to `flux mini batch` since it was trivial.

**Comment by garlick** [comment]
Couple of docs items even though this is still a WIP (just in case useful):

1) This `flux mini batch` help output might need to explain "instance":

> Submit a batch SCRIPT and ARGS to be run as the initial program of an
instance. If no batch script is provided, one will be read from stdin.

Maybe "initial program of a new Flux instance."?

2) There's nice coverage in `flux-mini(1)` of `batch` but none yet for `alloc`.

**Comment by grondo** [comment]
> There's nice coverage in flux-mini(1) of batch but none yet for alloc.

That should be there now.

**Comment by grondo** [comment]
> Couple of docs items even though this is still a WIP (just in case useful):

> This flux mini batch help output might need to explain "instance":

This _is_ helpful. I'm about to remove WIP. I liked your addition.

**Comment by grondo** [comment]
There's a problem with the changes in `flux-mini.py` though, so further refactoring may be needed.

**Comment by grondo** [comment]
> There's a problem with the changes in flux-mini.py though, so further refactoring may be needed.

The problem is that LGTM.com is calling out 
```
__init__ method calls overridden method
```
See [here](https://help.semmle.com/wiki/display/PYTHON/__init__+method+calls+overridden+method) for the description of this error.

However, I believe I've seen lots of examples in python of calling object methods from `__init__`, is this really a bug? It is very convenient behavior in this case, since `AllocCmd` can inherit from `BatchCmd`, but via a `post_init()` method, they can each add different positional argparse options to the end of the shared options.

Perhaps now that there are more than just two subcommands sharing options, options should just be built from functions and not within the `__init__` constructors of the subcommands. This would allow easier composition anyhow.. but it will be a delay in this PR.

**Comment by grondo** [comment]
Ok, based on coffee call conversation, I'll abstract out the common initialization from `BatchCmd` into an abstract class, then `BatchCmd` and `AllocCmd` can both inherit from that. @SteVwonder suggested in the future we may want to use argparse parent parser support to more cleanly allow composition of subcommands

https://docs.python.org/3/library/argparse.html#parents

**Comment by grondo** [comment]
> I'll abstract out the common initialization from BatchCmd into an abstract class, then BatchCmd and AllocCmd can both inherit from that.

Actually, it seemed a lot more straightforward to pull out the common options into a function. This gives the subcommands more flexibility in which order they insert arguments, i.e. we can keep the allocation related options at the end of the `--help` listing.

If this seems ok (and no new errors pop out), I'll need to rework all the `flux-mini.py` commits in a more logical order.

**Comment by grondo** [comment]
Rebased to fix some spelling errors, and add doc changes as suggested by @garlick.

**Comment by grondo** [comment]
Ok, I've squashed everything up into nice(r) commits, added some more tests, and slightly enhanced documentation.

I've removed the WIP label now. A few people should try this out before we merge it.

**Comment by jameshcorbett** [comment]
> A few people should try this out before we merge it.

I will happily incorporate this into our code base once there's a version deployed on the TOSS machines. I know exactly what I would do with it, and I think it would be a pretty significant code-quality improvement for us. After going through several iterations of confusion, I actually think `flux mini batch` is exactly what the UQP needs for creating nested instances (for interactive use it's better than `bsub -Is/-I/-Ix` but I still think it's hard to beat `salloc -N1` and `lalloc 1`).

> I don't follow why a workflow tool would want to treat MPI programs the same as a new Flux instance.

The UQP treats MPI programs and new Flux instances equally in the same way that Flux does---i.e. it's just another job with some set of resources. However, the UQP keeps a flag associated with each job to determine whether it is actually an MPI program or a new Flux instance (because, like you said, in the end they're quite different). Before, this flag would determine whether the UQP executes `flux mini run` or `flux mini run -NX -nX ... flux start`. Now we can just swap out `flux mini run` for `flux mini batch`, keeping all of the `-n`, `-c`, and `-g` options the same. I think that's as convenient as it could be, and I don't think we have the need for some kind of automatic job-type deduction.

> E.g. you could write a script that passes through args to flux mini submit unless the string "flux start" appears in the commandline, in which case it passes through to flux mini batch.

If the string "flux start" appears at the command line, though, you would think the user/script already _knows_ that what they're executing is a new instance. I don't think there's any way to abstract that knowledge away from the user.

**Comment by grondo** [comment]
Thanks! I think one more change is needed here. `flux mini alloc` should only set `pty` shell option if stdin is a tty. I'll make that one-liner change and squash if that is alright.

**Comment by grondo** [comment]
Ok, force pushed a fix for `pty` and added a test to ensure `flux mini alloc` works when stdin is not a tty.

If tests pass, ok to set MWP?

**Comment by garlick** [comment]
> If tests pass, ok to set MWP?

Yes please!  (I responded with a thumbs up but probably should have given you a response that generates a notification :-)

**Comment by grondo** [comment]
Thanks!
